packages/kiota_serialization_form/lib/src/form_serialization_writer.dart

```diff
@@ -0,0 +1,220 @@
+part of '../kiota_serialization_form.dart';
+
+/// Represents a serialization writer that can be used to write a form url
+/// encoded string.
+class FormSerializationWriter implements SerializationWriter {
+  final StringBuffer _buffer = StringBuffer();
+  bool writingObject = false;
+
+  @override
+  ParsableHook? onAfterObjectSerialization;
+
+  @override
+  ParsableHook? onBeforeObjectSerialization;
+
+  @override
+  void Function(Parsable p, SerializationWriter w)? onStartObjectSerialization;
+
+  @override
+  Uint8List getSerializedContent() {
+    return utf8.encode(_buffer.toString());
+  }
+
+  @override
+  void writeAdditionalData(Map<String, dynamic> value) {
+    for (final entry in value.entries) {
+      _writeAnyValue(entry.key, entry.value);
+    }
+  }
+
+  void _writeAnyValue(String? key, Object? value) {
+    switch (value) {
+      case null:
+        writeNullValue(key);
+      case final bool b:
+        writeBoolValue(key, value: b);
+      case final int i:
+        writeIntValue(key, i);
+      case final double d:
+        writeDoubleValue(key, d);
+      case final String s:
+        writeStringValue(key, s);
+      case final DateTime d:
+        writeDateTimeValue(key, d);
+      case final DateOnly d:
+        writeDateOnlyValue(key, d);
+      case final TimeOnly t:
+        writeTimeOnlyValue(key, t);
+      case final Duration d:
+        writeDurationValue(key, d);
+      case final UuidValue u:
+        writeUuidValue(key, u);
+      default:
+        writeStringValue(key, value.toString());
+    }
+  }
+
+  @override
+  void writeBoolValue(String? key, {bool? value}) {
+    writeStringValue(key, value?.toString().toLowerCase());
+  }
+
+  @override
+  void writeByteArrayValue(String? key, Uint8List? value) {
+    if (value != null) {
+      writeStringValue(key, base64Encode(value));
+    }
+  }
+
+  @override
+  void writeCollectionOfEnumValues<T extends Enum>(
+    String? key,
+    Iterable<T>? values,
```

Comment:
I see that you have added a parsing function.
Should we also add a serialization function to map the code symbol to the wire value?

Reply:
We could, yes. I didn't because I think we should rely on language features as much as possible. The only thing missing in Dart is the ability to parse Strings back to the enum case in a generic way. Parsing it back from a concrete type is easy and must be provided by kiota.

Reply:
To avoid reflection in some languages we also generate the serialization method for enums which is in charge of translating a single enum value to its corresponding string representation. We could/should do the same in dart.

Reply:
Fair. I'll add corresponding signatures and update the tests.

Reply:
Here is what I think is the most complex example because of the language semantics.
I'm not sure we need all of that in dart.
The [string method](https://github.com/microsoftgraph/msgraph-sdk-go/blob/9692d59f6da0cf48261208887b5310752bbf4aa5/models/body_type.go#L9) is [used for single valued enums](https://github.com/microsoftgraph/msgraph-sdk-go/blob/9692d59f6da0cf48261208887b5310752bbf4aa5/models/item_body.go#L124) (non flaggable)
Otherwise, the [serialize method](https://github.com/microsoftgraph/msgraph-sdk-go/blob/9692d59f6da0cf48261208887b5310752bbf4aa5/models/body_type.go#L24) is used.

It could have been simpler if go had generics and supported flag enums at the time we originally implemented that.

Reply:
I'm not sure I understand... I hadn't anticipated flaggable enums yet and am not sure how to incorporate it (Dart doesn't support using enums as flags).

My current solution only offers passing a callback that will turn an enum value to a string. How do I need to change this to comply with other languages?

Reply:
This might be me getting confused by dart again. But how does this prototype accept any callback to serialize the enum values?

```dart
void writeCollectionOfEnumValues<T extends Enum>(
    String? key,
    Iterable<T>? values,
)
```

Reply:
It doesn't. In 25ae786c5ef2177c9418ac4fc08c5415b90e2e4a I added this change:

```patch
    void writeCollectionOfEnumValues<T extends Enum>(
      String? key,
      Iterable<T>? values,
+     EnumSerializer<T> serializer,
    );
```

The `serializer` is used for every value in the `values` to convert them to strings.

Reply:
(btw I am glad we are discussing all these details now before I delve deeper into implementing JSON and multipart! Thanks for being patient and helping with this effort @baywet)

Reply:
Thank you! Both for all the work and taking the time to ask all those questions ahead of implementation! This goes a long way in terms of quality end result and saving time to get there!
Resolving.